File: mostcommonword.js
//Objective is to find the most common word in a sentence in a string.
//There is also a list of banned words that cannot be chosen.

let paragraph = "Bob hit a ball, the hit BALL flew far after it was hit."
let banned  = ['hit']


//O(n) solution that uses regex as well as a hashmap to keep track of
//the frequencies of the words, as well as to remove any punctuation.

//Regex to get rid of any non-alphanumeric characters
let space = paragraph.toLowerCase().split(/\W+/g)
let map = {}

//Mapping the frequencies of each word
for (let word of space) {
    if (map[word] === undefined) {
        map[word] = 1
    } else {
        map[word]++
    }
}

let sorted = Object.keys(map).sort((a,b) => {
    return map[b] - map[a]
})

//Get rid of any banned words
let filter = sorted.filter(word => !banned.includes(word))
return filter[0]

File: mostcommonword.java
//Java Solution

class Solution {
    public String mostCommonWord(String paragraph, String[] banned) {
        
        String normalizedStr = paragraph.replaceAll("[^a-zA-Z0-9 ]", " ").toLowerCase();
        String[] words = normalizedStr.split("\\s+");
        
        Set<String> set = new HashSet();
        for (String word : banned) {
            set.add(word);
        }
        
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            if (!set.contains(word)) {
                map.put(word, map.getOrDefault(word, 0) + 1);
            }
        }
        
        return Collections.max(map.entrySet(), Map.Entry.comparingByValue()).getKey();
    }
}